politeiawww/politeiawww.go

```diff
@@ -858,6 +866,95 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
 	}
 }
 
+// updateBestBlock updates the cached best block.
+func (p *politeiawww) updateBestBlock(bestBlock uint64) {
+	p.bbMtx.Lock()
+	defer p.bbMtx.Unlock()
+	p.bestBlock = bestBlock
+}
+
+// getBestBlock returns the cached best block if there is an active websocket
+// connection to dcrdata. Otherwise, it requests the best block from politeiad
+// using the the decred plugin best block command.
+func (p *politeiawww) getBestBlock() (uint64, error) {
+	p.bbMtx.RLock()
+	bb := p.bestBlock
+	p.bbMtx.RUnlock()
+
+	// the cached best block will equal 0 if there is no active websocket
+	// connection to dcrdata, or if no new block messages have been received
+	// since a connection was established.
+	if bb == 0 {
+		return p.getBestBlockDecredPlugin()
+	}
+
+	return bb, nil
+}
+
+// resetPiDcrdataWSSubs is responsible for resetting the wsDcrdata connection
+// and making necessary changes to the required changes to the politeiawww
+// state so that the service continues to function properly during and after
+// the reconnection.
+func (p *politeiawww) resetPiDcrdataWSSubs() error {
+	// The cached best block is set to zero so that in the time between
+	// reconnection and receiving the first new block message, instead of
+	// using the old cached value, politeiad is queried for the best block.
+	p.updateBestBlock(0)
+
+	return p.wsDcrdata.reconnect()
+}
+
+// setupPiDcrdataWSSubs subscribes and listens to websocket messages from
+// dcrdata that are needed for pi.
+func (p *politeiawww) setupPiDcrdataWSSubs() error {
+	err := p.wsDcrdata.subToNewBlock()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+	listenerLoop:
+		for {
+			msg, ok := <-p.wsDcrdata.Receive()
+			if !ok {
+				log.Errorf("Error receiving msg from dcrdata, msg: %v", msg)
+				err = p.resetPiDcrdataWSSubs()
+				if err != nil {
+					log.Errorf("Unable to reconnect to dcrdata, no longer " +
+						"listening.")
+					break listenerLoop
+				}
+				continue
+			}
+
+			switch m := msg.Message.(type) {
+			case *exptypes.WebsocketBlock:
+				log.Debugf("wsDcrdata message WebsocketBlock(height=%s)",
+					m.Block.Height)
+				p.updateBestBlock(uint64(m.Block.Height))
+			case *pstypes.HangUp:
```

Comment:
The current pubsub server will also be sending ping signals (`pstypes.SigPingAndUserCount`) at a regular interval.  Might as well catch that so it doesn't end up in the error logs.

Reply:
To expand on this, the pings were supposed to be going out from the server -> clients, but there was a bug.  With the fix in, the client can/should set their read timeout appropriately (at least `pubsub.PingInterval`), here: https://github.com/decred/politeia/pull/1021/files#diff-ec12e4d6d15a45b1c30aa86a9f67b420R192.  Please have a look at the dcrdata PR that fixes the ping and exports the server ping interval: https://github.com/decred/dcrdata/pull/1645/files

If I understand the current read timeout strategy in this PR, it is to reconnect.  Ideally that would not be necessary as regular pings will be coming through so the client knows the connection is healthy.

Reply:
Oh, sorry I forget that read timeouts in psclient don't raise to an error in the consumer.  Although a big change in https://github.com/decred/dcrdata/pull/1645/files is that legit read timeouts will be errors since the server sent ping should be relied upon for health.

I'd suggest updating the the pubsub dependency after I merge https://github.com/decred/dcrdata/pull/1645 (today) and set the read timeout to `pubsub.PingInterval + something` or use the default read timeout va nil Opts which sets it to `pubsub.PingInterval * 10 / 9`.

Reply:
The server updates are merged and deployed.  Now you'd just need to update the pubsub import and set the read timeout appropriately.  e.g. https://github.com/decred/politeia/commit/53ee8215937fbbe6ed9db37bca92220867922619
You can cherry-pick 53ee8215937fbbe6ed9db37bca92220867922619 if it works for you.

Reply:
You'll also want that empty `case *pstypes.SigPingAndUserCount:` too.

Reply:
@chappjc Sounds good, but could you also deploy to the testnet dcrdata? That one is still using 3.1.0.

Reply:
Now deployed to testnet.dcrdata.org.

Reply:
Thank you sir.